projects/ngneat/dialog/src/lib/types.ts

```diff
@@ -24,3 +25,40 @@ export interface DialogContentData {
 export type JustProps<T extends object> = {
   [K in keyof T]: T[K] extends (...args: any[]) => any ? never : T[K];
 };
+
+export type ExtractDialogRefData<T extends DialogRef> = T extends DialogRef<infer Data, any, any> ? Data : never;
+
+type _AnyIfUnknown<T> = unknown extends T ? any : T;
+
+export type ExtractDialogRefResult<T extends DialogRef> = T extends DialogRef<any, infer Result, any>
+  ? _AnyIfUnknown<Result>
+  : never;
+
+export type ExtractDialogRefType<T extends DialogRef> = T extends DialogRef<any, any, infer Type> ? Type : never;
+
+type _ExtractComponentDialogRef<T extends Type<unknown>> = T extends Type<unknown>
+  ? TupleExtract<ConstructorParameters<T>, DialogRef<unknown, unknown>>
+  : never;
+
+type ExtractComponentDialogRef<T extends Type<unknown>> = _ExtractComponentDialogRef<T> extends [infer U]
+  ? U
+  : // Any is necessary to not break things: injecting DialogRef in the component constructor is optional.
+    any;
+
+export type ComputedDialogRefType<T extends Type<unknown> | TemplateRef<unknown>> = T extends TemplateRef<unknown>
+  ? any
```

Comment:
This is not typed voluntarily at the moment, is the generic of TemplateRef<T> the Result param of DialogRef? If so, I think it's possible to handle this case to have something like this:

```typescript
export function open(template: TemplateRef<Output>) {
  const ref = this.dialog.open(template);
  // "data" parameter is Output type
  ref.afterClosed$.subscribe(data => [...])
}
```